### PR TITLE
GOVSI-642: add reset password request lambda

### PR DESCRIPTION
**Pull request by dbes-gds**
## What?

Add reset password request lambda.

Sends a reset password email with a one-time link to allow the user to change their password.

## Why?

User needs to be able to reset their password (if they have forgotten it, for example)

## Related PRs

#533

### CODE REVIEW

File: ci/terraform/oidc/reset-password-request.tf [hunk 1]
```diff
@@ -0,0 +1,51 @@
+module "reset-password-request" {
```
Comment (JHjava) [inline ci/terraform/oidc/reset-password-request.tf:1]: You'll need to add the triggers and the depends_on to the oidc/api-gateway.tf

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    private APIGatewayProxyResponseEvent handleNotificationRequest(
+            String email,
+            NotificationType notificationType,
+            Session session,
+            SessionState nextState)
+            throws JsonProcessingException {
+
+        String subjectId = authenticationService.getSubjectFromEmail(email).getValue();
```
Comment (JHjava) [inline frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java:159]: Do we expect this to always find the user or is there a chance that a user might not exist with that email?

File: ci/terraform/oidc/sqs.tf [hunk 3]
```diff
@@ -134,6 +134,9 @@ resource "aws_lambda_function" "email_sqs_lambda" {
       VERIFY_EMAIL_TEMPLATE_ID        = "b7dbb02f-941b-4d72-ad64-84cbe5d77c2e"
       VERIFY_PHONE_NUMBER_TEMPLATE_ID = "7dd388f1-e029-4fe7-92ff-18496dcb53e9"
       MFA_SMS_TEMPLATE_ID             = "7dd388f1-e029-4fe7-92ff-18496dcb53e9"
+      RESET_PASSWORD_TEMPLATE_ID      = "0aaf3ae8-1825-4528-af95-3093eb13fda0"
+      FRONTEND_BASE_URL               = var.frontend_base_url
```
Comment (chrisclayson) [inline ci/terraform/oidc/sqs.tf:138]: Doesn't this need to be `local.frontend_base_url` as we added this morning?

File: echo-int-test-vars.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
Comment (chrisclayson) [inline echo-int-test-vars.sh:1]: Did you mean to commit this?

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java (hunk 2 above)
Comment (dbes-gds) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java:159]: They have to have gone through checkuserexists to get here so in theory the user should exist.  It's possible if someone is running two flows at the same time and deletes the user.  Could make it return an error?

File: ci/terraform/oidc/sqs.tf (hunk 3 above)
Comment (dbes-gds) [reply ci/terraform/oidc/sqs.tf:138]: good spot

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java (hunk 2 above)
Comment (JHjava) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/ResetPasswordRequestHandler.java:159]: This is a very rare edge case as the user would have to delete their account between the time they enter their email and clicking the check password link therefore it we can let it blow up if the user isn't there. No action needed

File: ci/terraform/oidc/reset-password-request.tf (hunk 1 above)
Comment (dbes-gds) [reply ci/terraform/oidc/reset-password-request.tf:1]: added

File: echo-int-test-vars.sh (hunk 4 above)
Comment (dbes-gds) [reply echo-int-test-vars.sh:1]: yes, it generates the run config string you can use to run integration tests in the ide

**Review by JHjava — APPROVED** [review]

**Pull request merged**